Support plugins without arguments

This currently errors when there's a plugin being referenced that doesn't take arguments, e.g.:

```
steps:
  - command: 'echo wow'
    plugins:
      - seek-oss/docker-ecr-cache#v0.0.1:
      - docker#v2.0.0:
```

Results in
```
2018/10/23 09:42:22 $ buildkite-agent pipeline upload --dry-run

2018-10-23 09:42:22 INFO   Searching for pipeline config...

2018-10-23 09:42:22 INFO   Found config file ".buildkite/pipeline.yml"

panic: interface conversion: interface {} is string, not map[string]interface {}



goroutine 1 [running]:

main.SharedSecretSigner.extractPlugins(0xc000406d50, 0x6, 0x0, 0x0, 0x852460, 0xc000410020, 0x878f40, 0xc0003d8230, 0x94, 0x10)

	/Users/zsims/projects/buildkite-signed-pipeline/signer.go:154 +0x5b5

main.SharedSecretSigner.signStep(0xc000406d50, 0x6, 0x0, 0x0, 0x878f40, 0xc000188000, 0x194, 0xc0003d81d0, 0x94, 0xc00011bac8, ...)

	/Users/zsims/projects/buildkite-signed-pipeline/signer.go:104 +0x984

main.SharedSecretSigner.Sign(0xc000406d50, 0x6, 0x0, 0x0, 0x87aa40, 0xc000074810, 0xc000000300, 0xc00011bc40, 0x81065c, 0xc00008c238)

	/Users/zsims/projects/buildkite-signed-pipeline/signer.go:57 +0x668

main.(*uploadCommand).run(0xc00007ce40, 0xc000185c20, 0x85b0e0, 0xc000073400)

	/Users/zsims/projects/buildkite-signed-pipeline/main.go:98 +0xf4

main.(*uploadCommand).run-fm(0xc000185c20, 0x409af3, 0x88f900)

	/Users/zsims/projects/buildkite-signed-pipeline/main.go:41 +0x34

gopkg.in/alecthomas/kingpin%2ev2.(*actionMixin).applyActions(0xc00018a0d8, 0xc000185c20, 0x0, 0x0)

	/Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/actions.go:28 +0x6d

gopkg.in/alecthomas/kingpin%2ev2.(*Application).applyActions(0xc0001ac690, 0xc000185c20, 0x0, 0x0)

	/Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/app.go:557 +0xdf

gopkg.in/alecthomas/kingpin%2ev2.(*Application).execute(0xc0001ac690, 0xc000185c20, 0xc0000733f0, 0x1, 0x1, 0x0, 0x0, 0xc0001a6960, 0x0)

	/Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/app.go:390 +0x8f

gopkg.in/alecthomas/kingpin%2ev2.(*Application).Parse(0xc0001ac690, 0xc00007c030, 0x1, 0x1, 0x1, 0xc00008c1d0, 0x0, 0x1)

	/Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/app.go:222 +0x1fa

main.main()

	/Users/zsims/projects/buildkite-signed-pipeline/main.go:79 +0x914

[31m🚨 Error: The command exited with status 2[0m

```

--- a/plugins.go
+++ b/plugins.go
@@ -17,6 +17,23 @@ type Plugin struct {
 	Params map[string]interface{}
 }
 
+func NewPluginFromReference(item interface{}) (*Plugin, error) {
+	switch item.(type) {
+	// plugin references that are just a plugin name, e.g. docker#v1.2.3
+	case string:
+		return &Plugin{item.(string), nil}, nil
+	// plugin references that are a name and a set of settings
+	case map[string]interface{}:
+		for name, settings := range item.(map[string]interface{}) {
+			// note that x.(T) is avoided here as settings may be null in the case
+			// of plugins without parameters
+			parameters, _ := settings.(map[string]interface{})
+			return &Plugin{name, parameters}, nil
+		}
+	}
+	return nil, fmt.Errorf("Unknown plugin reference type %T", item)
+}
+
 func (p Plugin) Repository() string {
 	if m := officialPluginRegex.FindStringSubmatch(p.Name); len(m) == 3 {
 		return fmt.Sprintf(`github.com/buildkite-plugins/%s-buildkite-plugin%s`, m[1], m[2])

--- a/plugins.go
+++ b/plugins.go
@@ -18,13 +18,13 @@ type Plugin struct {
 }
 
 func NewPluginFromReference(item interface{}) (*Plugin, error) {
-	switch item.(type) {
+	switch i := item.(type) {
 	// plugin references that are just a plugin name, e.g. docker#v1.2.3
 	case string:
-		return &Plugin{item.(string), nil}, nil
+		return &Plugin{i, nil}, nil
 	// plugin references that are a name and a set of settings
 	case map[string]interface{}:
-		for name, settings := range item.(map[string]interface{}) {
+		for name, settings := range i {
 			// note that x.(T) is avoided here as settings may be null in the case
 			// of plugins without parameters
 			parameters, _ := settings.(map[string]interface{})

--- a/signer.go
+++ b/signer.go
@@ -149,15 +149,36 @@ func (s SharedSecretSigner) extractPlugins(plugins interface{}) (string, error) 
 	var parsed []Plugin
 
 	switch t := plugins.(type) {
+	/*
+	 handles array syntax for referencing plugins, e.g.
+	 plugins:
+	  - foo#v1.2.3
+	  - bar#v1.2.3:
+	  - another#v1.2.3:
+	    a-parameter: true
+	*/
 	case []interface{}:
 		for _, item := range t {
-			for name, settings := range item.(map[string]interface{}) {
-				parsed = append(parsed, Plugin{name, settings.(map[string]interface{})})
+			plugin, err := NewPluginFromReference(item)
+			if err != nil {
+				return "", err
 			}
+			parsed = append(parsed, *plugin)
 		}
+	/*
+	 handles object syntax for referencing plugins, e.g.
+	 plugins:
+	  bar#v1.2.3:
+	  another#v1.2.3:
+	    a-parameter: true
+	*/
 	case map[string]interface{}:
-		for name, settings := range t {
-			parsed = append(parsed, Plugin{name, settings.(map[string]interface{})})
+		for item := range t {
+			plugin, err := NewPluginFromReference(item)
+			if err != nil {
+				return "", err
+			}
+			parsed = append(parsed, *plugin)
 		}
 	default:
 		return "", fmt.Errorf("Unknown plugin type %T", t)

--- a/signer_test.go
+++ b/signer_test.go
@@ -121,6 +121,16 @@ func TestPipelines(t *testing.T) {
 			`{"steps":[{"env":{"STEP_SIGNATURE":"signature(,[{\"github.com/buildkite-plugins/docker-buildkite-plugin#v1.4.0\":{\"image\":\"node:7\"}}])"},"label":"I have no commands","plugins":[{"docker#v1.4.0":{"image":"node:7"}}]}]}`,
 		},
 		{
+			"Plugin object syntax without arguments",
+			`{"steps":[{"label":"I have no commands","plugins":[{"docker#v1.4.0":null}]}]}`,
+			`{"steps":[{"env":{"STEP_SIGNATURE":"signature(,[{\"github.com/buildkite-plugins/docker-buildkite-plugin#v1.4.0\":null}])"},"label":"I have no commands","plugins":[{"docker#v1.4.0":null}]}]}`,
+		},
+		{
+			"Plugin array syntax without arguments",
+			`{"steps":[{"label":"I have no commands","plugins":["docker#v1.4.0"]}]}`,
+			`{"steps":[{"env":{"STEP_SIGNATURE":"signature(,[{\"github.com/buildkite-plugins/docker-buildkite-plugin#v1.4.0\":null}])"},"label":"I have no commands","plugins":["docker#v1.4.0"]}]}`,
+		},
+		{
 			"Pipeline with multiple steps",
 			`{"steps":[{"command":"echo hello"},{"commands":["echo world", "echo foo"]}]}`,
 			`{"steps":[{"command":"echo hello","env":{"STEP_SIGNATURE":"signature(echo hello,)"}},{"commands":["echo world","echo foo"],"env":{"STEP_SIGNATURE":"signature(echo world\necho foo,)"}}]}`,

